sdk/eventhub/Azure.Messaging.EventHubs/src/Amqp/AmqpEventHubProducer.cs

```diff
@@ -0,0 +1,417 @@
+
+                        EventHubsEventSource.Log.EventPublishStart(EventHubName, logPartition, messageHash);
+
+                        link = await SendLink.GetOrCreateAsync(UseMinimum(ConnectionScope.SessionTimeout, _tryTimeout)).ConfigureAwait(false);
+                        cancellationToken.ThrowIfCancellationRequested<TaskCanceledException>();
+
+                        // Validate that the batch of messages is not too large to send.  This is done after the link is created to ensure
+                        // that the maximum message size is known, as it is dictated by the service using the link.
+
+                        if (batchMessage.SerializedMessageSize > MaximumMessageSize)
+                        {
+                            throw new MessageSizeExceededException(EventHubName, string.Format(Resources.MessageSizeExceeded, messageHash, batchMessage.SerializedMessageSize, MaximumMessageSize));
+                        }
+
+                        // Attempt to send the message batch.
+
+                        var deliveryTag = new ArraySegment<byte>(BitConverter.GetBytes(Interlocked.Increment(ref _deliveryCount)));
+                        var outcome = await link.SendMessageAsync(batchMessage, deliveryTag, AmqpConstants.NullBinary, _tryTimeout.CalculateRemaining(stopWatch.Elapsed)).ConfigureAwait(false);
+                        cancellationToken.ThrowIfCancellationRequested<TaskCanceledException>();
+
+                        if (outcome.DescriptorCode != Accepted.Code)
+                        {
+                            throw AmqpError.CreateExceptionForError((outcome as Rejected)?.Error, EventHubName);
+                        }
+
+                        // The send operation should be considered successful; return to
+                        // exit the retry loop.
+
+                        return;
+                    }
+                    catch (AmqpException amqpException)
+                    {
+                        throw AmqpError.CreateExceptionForError(amqpException.Error, EventHubName);
+                    }
+                    catch (Exception ex)
+                    {
+                        EventHubsEventSource.Log.EventPublishError(EventHubName, logPartition, messageHash, ex.Message);
+
+                        // Determine if there should be a retry for the next attempt; if so enforce the delay but do not quit the loop.
+                        // Otherwise, bubble the exception.
+
+                        ++failedAttemptCount;
+                        retryDelay = _retryPolicy.CalculateRetryDelay(ex, failedAttemptCount);
+
+                        if ((retryDelay.HasValue) && (!ConnectionScope.IsDisposed))
+                        {
+                            cancellationToken.ThrowIfCancellationRequested<TaskCanceledException>();
+
+                            await Task.Delay(retryDelay.Value, cancellationToken).ConfigureAwait(false);
```

Comment:
We usually log a `TaskCanceledException` when it's thrown in the `try` block. Shouldn't we do the same here? The way it is, it won't go through the `EventPublishError` call.

Reply:
Same question for the final `throw` in the end of the `catch` block.

Reply:
You're correct.    I did some restructuring around each of the operation retry loops in my AMQP refactoring pass. I'll defer this until that PR is out.